### PR TITLE
[JUJU-316] DQLite auto-cluster in ha mode with TLS support

### PR DESCRIPTION
**Pull request by achilleasa**
This PR refactors and enhances the previous implementation so that
we can automatically establish the dqlite cluster based on the
information that we can obtain from the agent's configuration.

In particular, the code will now leverage the (generated) CA and
controller certificates used by the controller to encrypt the traffic
between the dqlite instances. In addition, we rely on client certificate
authentication to prevent unauthorized clients from connecting to
dqlite.

Another aspect of the refactoring work has to do with the selection of a
suitable address for dqlite to listen to and advertise to other peers.
This is achieved by scanning the machine's NIC addresses and selecting
the first available cloud-local address. By avoiding public addresses we
ensure that DB traffic is always routed internally (important on public
clouds where public traffic may incur additional costs).

The last bit of the puzzle is the detection of usable remote peers. We
setup a TCP listener on a random port that serves a random nonce. Then
we iterate the list of API addresses, filter out any address that does
not hava a cloud-local scope (see above), extract the host part of the
IP and try to connect to the nonce port on that IP (read timeouts ensure
we bail out fast if the nonce server is not reachable) and read back the
nonce.

This allows auto-clustering to work because the agent configuration for
the HA nodes is guaranteed to contain the cloud-local API address for
the original controller. As a result, the newly added HA machines can
always reach the dqlite server on the original controller and discover
the other peers via RAFT.

## QA steps (x86_64 only)

```sh
# If you don't have musl installed make sure to follow the hints emitted by the Makefile errors!
$ make go-build cgo-go-build
$ _build/linux_amd64/bin/juju bootstrap lxd test --no-gui --build-agent
$ juju enable-ha

# Wait for the machines to join the HA cluster...

# Connect to machine-0 and check the logs (CTRL+D to exit)
$ make juju-dqlite-repl
.open logs
SELECT entity, COUNT(*) from logs group by entity;

# Open a new terminal, connect to the other machines and ensure 
# that you can still access all logs by running the same commands as above
# 
# Use lxc list to find the  names of the other boxes
$ JUJU_CONTROLLER_BOX=juju-XXX-1 make juju-dqlite-repl
.open logs
SELECT entity, COUNT(*) from logs group by entity;

$ JUJU_CONTROLLER_BOX=juju-XXX-2 make juju-dqlite-repl
.open logs
SELECT entity, COUNT(*) from logs group by entity;

# Keep running the same queries on the different terminals and verify that the logs are in sync.
```

### CODE REVIEW

File: worker/dbaccessor/package_test.go [hunk 1]
```diff
@@ -1,14 +0,0 @@
...
-
-import (
-	"testing"
-
-	gc "gopkg.in/check.v1"
-)
-
-func TestPackage(t *testing.T) {
-	gc.TestingT(t)
-}
```
Comment (SimonRichardson) [inline worker/dbaccessor/package_test.go:14]: We can still just keep this file, even if the tests don't exist _yet_

Comment (achilleasa) [reply worker/dbaccessor/package_test.go:14]: I'd rather we bring it back when we have some actual tests here.

File: worker/dbaccessor/worker.go [hunk 2]
```diff
@@ -238,11 +271,64 @@ func (w *dbWorker) initializeDqlite() error {
...
+	// peers. To figure out the address to bind to, we need to scan all NIC
+	// addresses and return the first one that has cloud-local scope.
+	for _, sysAddr := range sysAddrs {
+		var host string
+
+		switch v := sysAddr.(type) {
+		case *net.IPNet:
+			host = v.IP.String()
+		case *net.IPAddr:
+			host = v.IP.String()
```
Comment (juanmanuel-tirado) [inline worker/dbaccessor/worker.go:313]: default case?

**Review by juanmanuel-tirado — APPROVED** [review]
Proceed after minor changes. Nice work!

File: worker/dbaccessor/worker.go (hunk 2 above)
Comment (achilleasa) [reply worker/dbaccessor/worker.go:313]: It's caught by the if block immediately below the switch.

**Pull request merged**